snuba/admin/jwt.py

```diff
@@ -0,0 +1,68 @@
+from typing import Any, Optional
+
+import structlog
+
+from snuba.admin.user import AdminUser
+
+logger = structlog.get_logger().bind(module=__name__)
+
+
+CERTS: Optional[Any] = None
+AUDIENCE: Optional[str] = None
+
+
+def _certs() -> Any:
+    """Returns a dictionary of current Google public key certificates for
+    validating Google-signed JWTs. Since these change rarely, the result
+    is cached on first request for faster subsequent responses.
+    """
+    import requests
+
+    global CERTS
+    if CERTS is None:
+        response = requests.get("https://www.gstatic.com/iap/verify/public_key")
+        CERTS = response.json()
+    return CERTS
+
+
+def _get_metadata(item_name: str) -> str:
+    """Returns a string with the project metadata value for the item_name.
+    See https://cloud.google.com/compute/docs/storing-retrieving-metadata for
+    possible item_name values.
+    """
+    import requests
+
+    endpoint = "http://metadata.google.internal"
+    path = "/computeMetadata/v1/project/"
+    path += item_name
+    response = requests.get(
```

Comment:
I am not good at auth: don't we need to pass in some headers or something so Google knows which project we are trying to access?

Reply:
My understanding is that this API returns the project ID based on the GKE cluster that's initiating the request.

From https://cloud.google.com/python/docs/getting-started/authenticate-users#create_the_source_code:

> The App Engine metadata service (and similar metadata services for other Google Cloud computing services) looks like a web site and is queried by standard web queries. However, it isn't actually an external site, but an internal feature that returns requested information about the running app, so it is safe to use http instead of https requests. It's used to get the current Google Cloud identifiers needed to define the JWT assertion's intended audience.

Reply:
Ah I see. Man Google is a smart company.